spec/lib/topic_query_spec.rb

```diff
@@ -2091,6 +2091,32 @@ def suggested_for(topic)
     end
   end
 
+  describe "#apply_ordering" do
+    fab!(:topic1) { Fabricate(:topic, spam_count: 3, bumped_at: 3.hours.ago) }
+    fab!(:topic2) { Fabricate(:topic, spam_count: 2, bumped_at: 3.minutes.ago) }
+    fab!(:topic3) { Fabricate(:topic, spam_count: 3, bumped_at: 1.hour.ago) }
+
+    it "applies both topic_query_sortable_mapping and topic_query_apply_ordering_result modifiers" do
+      plugin_instance = Plugin::Instance.new
+
+      plugin_instance.register_modifier(
+        :topic_query_sortable_mapping,
+      ) do |sortable_mapping, topic_qeury|
+        sortable_mapping["spam"] = "spam_count"
+        sortable_mapping
+      end
+
+      plugin_instance.register_modifier(
+        :topic_query_apply_ordering_result,
+      ) do |result, sort_column, sort_dir, options, topic_query|
+        result.order("topics.#{sort_column} #{sort_dir}, bumped_at DESC")
+      end
```

Comment:
Having to use two modifiers just for this feature to work is not ideal and can be confusing for developers that are trying to use it. Given the fact that we do not document plugin modifiers generally, I think we need to look into simplifying this change to only use one modifier.

Reply:
Do also note that you have to [unregister the modifier](https://github.com/discourse/discourse/blob/2abf0f483c5a11820942f8dcc942a4bb9cfc3142/lib/discourse_plugin_registry.rb#L277) in order to avoid leaking state.

Reply:
I tried things out locally and this will likely work:

```
  def apply_ordering(result, options = {})
    order_option = options[:order]
    sort_dir = (options[:ascending] == "true") ? "ASC" : "DESC"

    new_result =
      DiscoursePluginRegistry.apply_modifier(
        :topic_query_apply_ordering_result,
        result,
        order_option,
        sort_dir,
        options,
        self,
      )

    return new_result if !new_result.nil? && new_result != result
    sort_column = SORTABLE_MAPPING[order_option] || "default"

    # If we are sorting in the default order desc, we should consider including pinned
    # topics. Otherwise, just use bumped_at.
    if sort_column == "default"
      if sort_dir == "DESC"
        # If something requires a custom order, for example "unread" which sorts the least read
        # to the top, do nothing
        return result if options[:unordered]
      end
      sort_column = "bumped_at"
    end

    # If we are sorting by category, actually use the name
    if sort_column == "category_id"
      # TODO forces a table scan, slow
      return result.references(:categories).order(<<~SQL)
        CASE WHEN categories.id = #{SiteSetting.uncategorized_category_id.to_i} THEN '' ELSE categories.name END #{sort_dir}
      SQL
    end

    if sort_column == "op_likes"
      return(
        result.includes(:first_post).order(
          "(SELECT like_count FROM posts p3 WHERE p3.topic_id = topics.id AND p3.post_number = 1) #{sort_dir}",
        )
      )
    end

    if sort_column.start_with?("custom_fields")
      field = sort_column.split(".")[1]
      return(
        result.order(
          "(SELECT CASE WHEN EXISTS (SELECT true FROM topic_custom_fields tcf WHERE tcf.topic_id::integer = topics.id::integer AND tcf.name = '#{field}') THEN (SELECT value::integer FROM topic_custom_fields tcf WHERE tcf.topic_id::integer = topics.id::integer AND tcf.name = '#{field}') ELSE 0 END) #{sort_dir}",
        )
      )
    end

    result = result.order("topics.#{sort_column} #{sort_dir}")
  end
```

Reply:
Yes, this is working perfectly. I updated the spec tests according to that. Thank you! 👍